### PR TITLE
Transition from Collection to Dictionary

### PR DESCRIPTION
**Pull request by Crowes**
Inspired by `Collection` (in utils), [I rewrote it into a library](https://github.com/Sugarcoated/Fondant/wiki/Collection) that I basically use in all of my projects, and regularly keep a suite of tests updated for. I also went on to write [a class called `Dictionary`](https://github.com/Sugarcoated/Fondant/wiki/Dictionary) that inherits from `Collection`, intending to restrict the contents to instances of a particular `class` -- even handling base JavaScript classes such as `String` and `Number` via casting.

Discord.js, for the most part, relies heavily on custom structures wrapped around the Discord API responses. Implementing `Dictionary` provides an alternate use case to `Collection`, and in the case of Discord.js, this means you can restrict/predict what type of set of items your functions are going to return, while using `Collection` when your sets of items are different `class`es from each other.

This would require a relatively simple implementation. `Collection` has all the methods used in the library, and implements further methods that users could use when dealing with their bots. However, there are two concerns with implementing this:

1. It can't be done all at once, there are at least 260 uses of the word `Collection` throughout the library.
2. I'm only just getting my head around how the library works, so I'm not always sure what is expected from a method.

I've provided an example in my commit for usage.

**Semantic versioning classification:**  
- [ ] This PR changes the library's interface (methods or parameters added)
- [ ] This PR includes breaking changes (methods removed or renamed, parameters moved or removed)
- [ ] This PR **only** includes non-code changes, like changes to documentation, README, etc.

### CODE REVIEW

**Comment by devsnek** [comment]
-1 on this, this doesn't seem worth the dep and overhead

**Comment by kyranet** [comment]
I will be honest, I have checked the source code and it's giant. For the purpose we use `Collection` in discord.js and frameworks, is more than enough, reaching the point we rarely use the 100% of it.

`Collection` is widely used by the library, therefore, it's much better if it was maintained inside the library itself instead of fragmenting its utils into more libraries.

> I also went on to write a class called Dictionary that inherits from Collection, intending to restrict the contents to instances of a particular class -- even handling base JavaScript classes such as String and Number via casting.

That's what [DataStore](https://discord.js.org/#/docs/main/master/class/DataStore)s are for, to enforce its entries to be strictly instanceof its holds. The structure it has is easy to read, to understand, and it's very scalable. **In my opinion**, you're replicating the DataStore's functionality with something named "Dictionary".

Not to mention you also have an [issue](https://github.com/Sugarcoated/Fondant/issues/1), opened by yourself, that leds to a serious problem: `typeof new String('');` returns `'object'` where `typeof '';` returns `string`.

Now for code style, your `Collection` class (which `Dictionary` extends) has a weird naming, i.e. `Collection#wait` to iterate over keys but with asynchronous calls. To be honest, that job is what I do manually with a for-of.

RAM usage, I saw you cache much more data, while we have the privates `_array` and `_keyArray` consisting on `Array<[K, V]>` and `Array<K>` respectively, yours has three: `_collectionItems`, `_collectionKeys` and `_collectionValues`.

CPU usage, instead of being more "low level" (the developer knows if the entry exists or not), yours is very "high level" and checks many things, that leds to performance decrease, for example [Collection#delete()](https://github.com/Sugarcoated/Fondant/blob/dd7c717b001982ba0617a75b2fa9507d82e7a2ff/src/Globals/Collection.js#L147-L158) and [Collection#set()](https://github.com/Sugarcoated/Fondant/blob/dd7c717b001982ba0617a75b2fa9507d82e7a2ff/src/Globals/Collection.js#L125-L136).

In the bright side (I'm not going to be fully negative), yours has some interesting features such as [Collection#index](https://github.com/Sugarcoated/Fondant/blob/dd7c717b001982ba0617a75b2fa9507d82e7a2ff/src/Globals/Collection.js#L168-L185) (even thought Map isn't indexed, and the keys are actually the indexes, the "order" of keys is by the time they got `set`d: just like a regular object).

**Comment by Crowes** [comment]
I'll reply in the morning to this. It's late.

**Comment by Crowes** [comment]
Thanks for the great response, it gives me something to take issue with! :)

(p.s I've bolded the key points so you don't have to read it all.)

> I will be honest, I have checked the source code and it's giant. For the purpose we use Collection in discord.js and frameworks, is more than enough, reaching the point we rarely use the 100% of it.

Firstly, I'd like to point out that the `Collection` in `util/` is larger than mine with and without unnecessary comments. I did the math - and yes - there are more lines, but you have 6,424 more characters if I remove those comments, and 3,404 more if I don't. The current implementation (in a lot of cases) does more, or the same thing, to get the result.

**For the sake of the dependency, you'd be reducing the weight of the file. And you using  it isn't the only consideration that should go into implementations, because your users are also exposed to it.**


> Collection is widely used by the library, therefore, it's much better if it was maintained inside the library itself instead of fragmenting its utils into more libraries.

This is an entirely valid point. While I could point out an example of all the dependencies that are maintained outside the library, but I understand the whole *not reinventing the wheel* thing.

This is not me saying that you are doing this by not using the dependency, just that you have your own implementation so it's either re-implementation or abstraction. **Considering the only breaking thing about this dependency is any methods I decided to remove/move. Everything else is still there.**

> I also went on to write a class called Dictionary that inherits from Collection, intending to restrict the contents to instances of a particular class -- even handling base JavaScript classes such as String and Number via casting.

> That's what DataStores are for, to enforce its entries to be strictly instanceof its holds. The structure it has is easy to read, to understand, and it's very scalable. In my opinion, you're replicating the DataStore's functionality with something named "Dictionary".

This actually does show potential for breaking changes. I can't gather what DataStore is doing in the source other than what you have explained. Just wanted to mention that **in some way, yes** (although I did not know about `DataStore` at time of writing it)**, but `Dictionary` overrides `Collection`'s methods for that instance safety, it does not implement its own**. I mentioned above that your users are also exposed to this, and users want to go, "Oh, it's just like `Map`", and use the `Map` prototype, not have to read the documentation to figure out they should be using `DataStore#add` and not an override of `Map#set`. 


> Not to mention you also have an issue, opened by yourself, that leds to a serious problem: typeof new String(''); returns 'object' where typeof ''; returns string.

This is because `Dictionary` is expected to *hold* a type, like `String`. Therefore [it is cast to that type if you use the `Dictionary#set` override](https://github.com/Sugarcoated/Fondant/blob/master/src/Globals/Dictionary.js#L96). However, that expectation is to be held only on contained items, not on those returned, and therefore it should be returning the primitive upon acccession. Something for me to fix in the meantime.

> Now for code style, your Collection class (which Dictionary extends) has a weird naming, i.e. Collection#wait to iterate over keys but with asynchronous calls. To be honest, that job is what I do manually with a for-of.

**There is nothing stopping the average joe using async/await and a for-of**. It was never in the release I wrote originally and **I was inspired to include it yesterday after learning there's no `await-able `#forEach`**. That's all it handles. You can still use `Map#forEach` if you don't care about there being a `Promise` returned that you can handle, just use the inherited method from `Map`. 

> RAM usage, I saw you cache much more data, while we have the privates _array and _keyArray consisting on Array<[K, V]> and Array<K> respectively, yours has three: _collectionItems, _collectionKeys and _collectionValues.

> CPU usage, instead of being more "low level" (the developer knows if the entry exists or not), yours is very "high level" and checks many things, that leds to performance decrease, for example Collection#delete() and Collection#set().

These two reasons are the only two I can't argue with! The whole caching thing was pretty much copied from your implementation, yet I must have missed this. **I'd assume you ignored a value cache due to redundancy, but it turns out that in my dependency its [`_collectionValues`] used far more when you just need to expose methods on the `Array` prototype than writing your own**.

> Many libraries, frameworks and bots that use Discord.js, even the ones that are very basic, rely heavily on Collection, that is, because they're useful, they have been there for enough long to be in many bots that are over a year old now, changing this will unfortunately break an entire network of libraries dependent from Discord.js' collections. (Did you see that even boilerplate bots that have very basic features other than a mini-command handler, cache the "commands" in a collection?).

But, as you've already covered `DataStore` is covering your needs as-is,; and as I've noticed, there are certain methods (I can agree) I've seen in boilerplate repositories too - `Collection#keyArray` on your implementation, being an example - and while what I'm proposing has some degree of potential, it's potential for breaking the library is far bigger than what it's worth. 

**TL;DR - I'm better using my knowledge to improve how your `Collection` works (maybe that #index thing!) and, even better, how `DataStore` works too should I firgure it out.**

**Comment by hydrabolt** [comment]
I agree with the points @kyranet has made, it would make more sense to keep Collection as part of discord.js if we rely on it so heavily.

Our DataStores are extensions of Collections that hold specific structures, such as messages and users. They implement their own creation/deletion/editing logic because some structures require more input than others.

Are there any specific, compelling reasons for us to use the Dictionary?

**Comment by Crowes** [comment]
> Are there any specific, compelling reasons for us to use the Dictionary?

Many, had this been proposed in 2015. Your `DataStore` is too ingrained in the library for it to be a simple switch to my Collection & Dictionary, even if I were to remove conflicts that @kyranet has mentioned. 

Case closed, frankly. Thanks for taking the time to discuss and whatnot.
